fetch.bs

```diff
@@ -1037,6 +1037,131 @@ worked on in <a href=https://github.com/whatwg/fetch/issues/1156>issue #1156</a>
+
+<p>To perform the <dfn>incrementally-read loop</dfn>, given a {{ReadableStreamDefaultReader}} object
+<var>reader</var>, <a>parallel queue</a> or <a for=/>global object</a> <var>taskDestination</var>,
+algorithm <var>processBodyChunk</var>, algorithm <var>processError</var>, and null or algorithm
+<var>processEndOfBody</var>:
+
+<ol>
+ <li>
+  <p>Let <var>readRequest</var> be the following <a>read request</a>:
+
+  <dl>
+   <dt><a for="read request">chunk steps</a>, given <var>chunk</var>
+   <dd>
+    <ol>
+     <li><p>Let <var>continueAlgorithm</var> be null.
+
+     <li><p>If <var>chunk</var> is not a {{Uint8Array}} object, then set
+     <var>continueAlgorithm</var> to this step: run <var>processError</var> given a {{TypeError}}.
+
+     <li>
+      <p>Otherwise:
+
+      <ol>
+       <li><p>Let <var>bytes</var> be a <a lt="get a copy of the buffer source">copy of</a>
+       <var>chunk</var>.
+
+       <li>
+        <p>Set <var>continueAlgorithm</var> to these steps:
+
+        <ol>
+         <li><p>Run <var>processBodyChunk</var> given <var>bytes</var>.
+
+         <li><p>Perform the <a>incrementally-read loop</a> given <var>reader</var>,
+         <var>taskDestination</var>, <var>processBodyChunk</var>, <var>processError</var>, and
+         <var>processEndOfBody</var>.
+        </ol>
+      </ol>
+
+     <li><p><a>Queue a fetch task</a> given <var>continueAlgorithm</var> and
+     <var>taskDestination</var>.
+    </ol>
+
+   <dt><a for="read request">close steps</a>
+   <dd><ol><li><p>If <var>processEndOfBody</var> is non-null, then <a>queue a fetch task</a> given
+   <var>processEndOfBody</var> and <var>taskDestination</var>.</ol>
+
+   <dt><a for="read request">error steps</a>, given <var>e</var>
+   <dd>
+    <ol>
+     <li><p>Let <var>errorAlgorithm</var> be this step: run <var>processError</var> given
+     <var>e</var>.
+
+     <li><p><a>Queue a fetch task</a> given <var>errorAlgorithm</var> and
+     <var>taskDestination</var>.
+    </ol>
+  </dl>
+
+ <li><p><a for=ReadableStreamDefaultReader>Read a chunk</a> from <var>reader</var> given
+ <var>readRequest</var>.
+</ol>
+
+<p>To <dfn export for=body>fully read</dfn> a <a for=/>body</a> <var>body</var>, given an algorithm
+<dfn export for="body/fully read"><var>processBody</var></dfn>, an optional algorithm
+<dfn export for="body/fully read"><var>processError</var></dfn>, and an optional null or
+<a for=/>global object</a> <dfn export for="body/fully read">taskDestination</dfn> (default null),
+run these steps:
+
+<ol>
+ <li><p>If <var>taskDestination</var> is null, then set <var>taskDestination</var> to the result of
+ <a>starting a new parallel queue</a>.
+
+ <li><p>Let <var>promise</var> be the result of <a>fully reading body as promise</a> given
+ <var>body</var>.
+
+ <li><p>Let <var>fulfilledSteps</var> given a <a>byte sequence</a> <var>bytes</var> be to
+ <a>queue a fetch task</a> to run <var>processBody</var> given <var>bytes</var>, with
+ <var>taskDestination</var>.
```

Comment:
@domenic for instance, in theory this shouldn't have to queue a task as there's no way for these fulfilled steps to run without one, but the streams primitives this builds on are not exactly clear about that.

This is a clear case, but I have similar questions about the incremental read above. Thoughts?

Apart from that I think this is largely the correct setup that we can use both for request (to incrementally read and then invoke transmit in the callback) and response bodies (mostly to read). And we can also use this to create a proxy stream for SRI, etc.

Reply:
fulfilledSteps and stream's chunk steps/close steps/error steps run as microtasks, on the stream object's relevant agent's event loop. So if you're happy with that event loop and that task priority, you don't need to queue a fetch task. You could do so if you want to switch event loops or change task priorities.

Reply:
Well, you've been asking for this functionality. What do you need? 😊 I could use some help here.

Reply:
Probably queuing a fetch task would be best, especially for HTML where a lot of the cases talk about the networking task source (e.g. "the last task that is queued by the networking task source once the resource has been fetched", "Queue an element task on the networking task source given el to process the linked resource", ...). It also seems more future-proof if we move to realm-independent "C++ streams".